challenge-manager/challenge-tree/generalized_path_timer_test.go

```diff
@@ -0,0 +1,523 @@
+		require.NoError(t, err)
+		wanted := HonestAncestors{
+			id("blk-4.a-5.a"),
+			id("blk-4.a-6.a"),
+			id("blk-4.a-8.a"),
+			id("blk-0.a-8.a"),
+			id("blk-0.a-16.a"),
+		}
+		for i, ancestor := range resp.AncestorEdgeIds {
+			if ancestor != wanted[i] {
+				t.Errorf("ancestor %d: wanted %s, got %s", i, wanted[i].Bytes(), ancestor.Bytes())
+			}
+		}
+	})
+	t.Run("big step challenge: many ancestors plus block challenge ancestors", func(t *testing.T) {
+		resp, err := tree.ComputeAncestorsWithTimers(ctx, id("big-5.a-6.a"), blockNum)
+		require.NoError(t, err)
+		wanted := HonestAncestors{
+			// Big step chal.
+			id("big-4.a-6.a"),
+			id("big-4.a-8.a"),
+			id("big-0.a-8.a"),
+			id("big-0.a-16.a"),
+			// Block chal.
+			id("blk-4.a-5.a"),
+			id("blk-4.a-6.a"),
+			id("blk-4.a-8.a"),
+			id("blk-0.a-8.a"),
+			id("blk-0.a-16.a"),
+		}
+		require.Equal(t, wanted, resp.AncestorEdgeIds)
+	})
+	t.Run("small step challenge: level zero edge has ancestors from big and block challenge", func(t *testing.T) {
+		resp, err := tree.ComputeAncestorsWithTimers(ctx, id("smol-0.a-16.a"), blockNum)
+		require.NoError(t, err)
+		wanted := HonestAncestors{
+			// Big step chal.
+			id("big-4.a-5.a"),
+			id("big-4.a-6.a"),
+			id("big-4.a-8.a"),
+			id("big-0.a-8.a"),
+			id("big-0.a-16.a"),
+			// Block chal.
+			id("blk-4.a-5.a"),
+			id("blk-4.a-6.a"),
+			id("blk-4.a-8.a"),
+			id("blk-0.a-8.a"),
+			id("blk-0.a-16.a"),
+		}
+		require.Equal(t, wanted, resp.AncestorEdgeIds)
+	})
+	t.Run("small step challenge: lowest level edge has full ancestry", func(t *testing.T) {
+		resp, err := tree.ComputeAncestorsWithTimers(ctx, id("smol-5.a-6.a"), blockNum)
+		require.NoError(t, err)
+		wanted := HonestAncestors{
+			// Small step chal.
+			id("smol-4.a-6.a"),
+			id("smol-4.a-8.a"),
+			id("smol-0.a-8.a"),
+			id("smol-0.a-16.a"),
+			// Big step chal.
+			id("big-4.a-5.a"),
+			id("big-4.a-6.a"),
+			id("big-4.a-8.a"),
+			id("big-0.a-8.a"),
+			id("big-0.a-16.a"),
+			// Block chal.
+			id("blk-4.a-5.a"),
+			id("blk-4.a-6.a"),
+			id("blk-4.a-8.a"),
+			id("blk-0.a-8.a"),
+			id("blk-0.a-16.a"),
+		}
+		require.Equal(t, wanted, resp.AncestorEdgeIds)
+	})
+}
+
+// The following tests checks a scenario where the honest
+// and dishonest parties take turns making challenge moves,
+// and as a result, their edges will be unrivaled for some time,
+// contributing to the path timer of edges we will query in this test.
+//
+// We first setup the following challenge tree, where branch `a` is honest.
+//
+//	 0-----4a----- 8a-------16a
+//		     \------8b-------16b
+//
+// Here are the creation times of each edge:
+//
+//	Alice (honest)
+//	  0-16a        = T1
+//	  0-8a, 8a-16a = T3
+//	  0-4a, 4a-8a  = T5
+//
+//	Bob (evil)
+//	  0-16b        = T2
+//	  0-8b, 8b-16b = T4
+//	  4a-8b        = T6
+//
+// In this contrived example, Alice and Bob's edges will have
+// a time interval of 1 in which they are unrivaled.
+func TestComputeHonestPathTimer(t *testing.T) {
```

Comment:
This same test already existed for the HonestPathTimer function that was hardcoded to 3 challenge levels. The same logic should pass, untouched, with this new implementatin